### PR TITLE
Exposing generic math support for unsigned right shift

### PR DESCRIPTION
**Pull request by tannergooding**

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Int16.cs [hunk 1]
```diff
@@ -1021,8 +1021,8 @@ public static bool TryCreate<TOther>(TOther value, out short result)
         /// <inheritdoc cref="IShiftOperators{TSelf, TResult}.op_RightShift(TSelf, int)" />
         static short IShiftOperators<short, short>.operator >>(short value, int shiftAmount) => (short)(value >> shiftAmount);
 
-        // /// <inheritdoc cref="IShiftOperators{TSelf, TResult}.op_UnsignedRightShift(TSelf, int)" />
-        // static short IShiftOperators<short, short>.operator >>>(short value, int shiftAmount) => (short)((ushort)value >> shiftAmount);
+        /// <inheritdoc cref="IShiftOperators{TSelf, TResult}.op_UnsignedRightShift(TSelf, int)" />
+        static short IShiftOperators<short, short>.operator >>>(short value, int shiftAmount) => (short)((ushort)value >>> shiftAmount);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Int16.cs:1025]: Why is the cast to ushort still needed here?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/IShiftOperators.cs [hunk 2]
```diff
@@ -22,11 +22,11 @@ public interface IShiftOperators<TSelf, TResult>
...
-        // /// <param name="shiftAmount">The amount by which <paramref name="value" /> is shifted right.</param>
-        // /// <returns>The result of shifting <paramref name="value" /> right by <paramref name="shiftAmount" />.</returns>
-        // /// <remarks>This operation is meant to perform n unsigned (otherwise known as a logical) right shift on all types.</remarks>
-        // static abstract TResult operator >>>(TSelf value, int shiftAmount); // TODO_GENERIC_MATH: shiftAmount should be TOther
+        /// <summary>Shifts a value right by a given amount.</summary>
+        /// <param name="value">The value which is shifted right by <paramref name="shiftAmount" />.</param>
+        /// <param name="shiftAmount">The amount by which <paramref name="value" /> is shifted right.</param>
+        /// <returns>The result of shifting <paramref name="value" /> right by <paramref name="shiftAmount" />.</returns>
+        /// <remarks>This operation is meant to perform n unsigned (otherwise known as a logical) right shift on all types.</remarks>
+        static abstract TResult operator >>>(TSelf value, int shiftAmount); // TODO_GENERIC_MATH: shiftAmount should be TOther
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/IShiftOperators.cs:30]: I assume there's an issue tracking this?

File: src/libraries/System.Private.CoreLib/src/System/Int16.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Int16.cs:1025]: C# doesn't define most operators for `byte`, `sbyte`, `short`, `ushort`, or `char` (outside the `op assignment` ones, like `+=` or `--`).

So in C#, `short >>> int` is actually `(int)short >>> int` which means `-1 >>> 1` goes from `0xFFFF` to `0xFFFF_FFFF` and then right shifts one giving `0x7FFF_FFFF`. Casting back to `short` then results in `0xFFFF` again.

Generic math returns and operates on the same type (like F# does) and so we need to explicitly treat it as a zero-extension so we get back `0x7FFF`

File: src/libraries/System.Private.CoreLib/src/System/Numerics/IShiftOperators.cs (hunk 2 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/IShiftOperators.cs:30]: The general design doc is tracking this at the moment. I'm just waiting for Aleksey to finish his PR and have the relevant changes stashed in a local branch

**Review by dakersnar — APPROVED** [review]
Looks good to me, besides the TODO Stephen mentioned.

**Comment by tannergooding** [comment]
@stephentoub let me know if you have any other feedback. Given you only had two minor questions, I've merged due to the deadline to get this in. Happy to address anything else in a follow up PR.

**Pull request merged**